### ISSUE TITLE
If autoload is off, projects cannot be exported until at least one is opened

### ISSUE THREAD

**Issue by ewpatton** [mit-cml/appinventor-sources#2419]
**Describe the bug**

A teacher reported that projects cannot be exported from the project list if autoload is off until at least one project has been opened.

**Affects**

- [ ] Designer
- [ ] Blocks editor
- [ ] Companion
- [ ] Compiled apps
- [ ] Buildserver
- [ ] Debugging
- [x] Other... (please describe)  Project list

**Expected behavior**

If autoload is turned off, checking a project and selecting "Export Project" should export the project. Instead, project ID 0 is requested and the server returns HTTP 500 (it should probably be a 403 or 404).

**Steps to reproduce**

1. In an account with projects, disable autoload via the settings menu
2. Reload App Inventor
3. Check the box next to a project
4. Use export project from the project menu
5. Note that no project is downloaded

Repeat the steps with the network tab open to observe the HTTP 500.

**Comment by Himanshu2107** [comment]
Hey, I would try to look into this. Please assign me this issue.

**Comment by Himanshu2107** [comment]
Can you point me to where can i find code related to "Export Project" and "autoload"?

**Comment by Himanshu2107** [comment]
I think I have found the bug.
First let us see what works and what does not work.

**When autoload is off and no project is selected**
1. If you select any number of projects from the list and then click on "Export selected project (.aia) to my computer", the download dialogue will not appear.
2. If you select "Export all projects" download dialogue will appear.

**When autoload is off and a project is selected**
 Both options work perfectly.

**When autoload is on**
Both options work perfectly.

When we click on the "Export selected project(.aia)" option the following snippet of code is called (in com/google/appinventor/client/TopToolbar.java):

```public void execute() {
      if (Ode.getInstance().getCurrentView() == Ode.PROJECTS) {
        List<Project> selectedProjects =
          ProjectListBox.getProjectListBox().getProjectList().getSelectedProjects();
        //If we are in the projects view
        if (selectedProjects.size() == 1) {
          exportProject(selectedProjects.get(0));
        } else if (selectedProjects.size() > 1) {
          exportSelectedProjects(selectedProjects);
        } else {
          // The user needs to select only one project.
          ErrorReporter.reportInfo(MESSAGES.wrongNumberProjectsSelected());
        }
      } else {
        //If we are in the designer view.
        Downloader.getInstance().download(ServerLayout.DOWNLOAD_SERVLET_BASE + ServerLayout.DOWNLOAD_PROJECT_SOURCE + "/" + Ode.getInstance().getCurrentYoungAndroidProjectId());
      }
    }
```

The bug is due to the first if condition ` if (Ode.getInstance().getCurrentView() == Ode.PROJECTS)`. 
When the Ode is initialized, the `currentView` is set to `DESIGNER`. 
(in com/google/appinventor/client/ode):
```// Remembers the current View
  public static final int DESIGNER = 0;
  public static final int PROJECTS = 1;
  public static final int USERADMIN = 2;
  public static final int TRASHCAN = 3;
  public static int currentView = DESIGNER;
```
So when autoload is off, we are in `PROJECTS` view but Ode incorrectly returns `DESIGNER` as the `currentView`. This causes the if condition to fail and the else clause
 `Downloader.getInstance().download(ServerLayout.DOWNLOAD_SERVLET_BASE + ServerLayout.DOWNLOAD_PROJECT_SOURCE + "/" + Ode.getInstance().getCurrentYoungAndroidProjectId());`
to run. This is run assuming we are in the designer section of appinventor. As we are not in the designer (and no project is selected), project-id is `0`. This causes the code in com/google/appinventor/client/utils/Downloader to ask the server for file(s) at link "http://localhost:8888/ode/download/project-source/0". This file(s) does not exist. Hence the server returns HTTP 500. 

If we rather select "Export all projects" we avoid the bug. 
```
private static class ExportAllProjectsAction implements Command {
    @Override
    public void execute() {
      Tracking.trackEvent(Tracking.PROJECT_EVENT,
          Tracking.PROJECT_ACTION_DOWNLOAD_ALL_PROJECTS_SOURCE_YA);

      // Is there a way to disable the Download All button until this completes?
      if (Window.confirm(MESSAGES.downloadAllAlert())) {

        Downloader.getInstance().download(ServerLayout.DOWNLOAD_SERVLET_BASE +
            ServerLayout.DOWNLOAD_ALL_PROJECTS_SOURCE);
      }
    }
```
As here it does not matter whether we are in `DESIGNER` or `PROJECTS` view.

Whenever a project is selected and we return to 'MyProjects' window it calls the following method:
```
public void switchToProjectsView() {
    // We may need to pass the code below as a runnable to
    // screenShotMaybe() so build the runnable now
    hideChaff();
    hideTutorials();
    Runnable next = new Runnable() {
        @Override
        public void run() {
          ProjectListBox.getProjectListBox().loadProjectList();
          currentView = PROJECTS;
          getTopToolbar().updateFileMenuButtons(currentView);
          deckPanel.showWidget(projectsTabIndex);
          // If we started a project, then the start button was disabled (to avoid
          // a second press while the new project wizard was starting (aka we "debounce"
          // the button). When the person switches to the projects list view again (here)
          // we re-enable it.
          projectToolbar.enableStartButton();
          projectToolbar.setProjectTabButtonsVisible(true);
          projectToolbar.setTrashTabButtonsVisible(false);
        }
      };
    if (designToolbar.getCurrentView() != DesignToolbar.View.BLOCKS) {
      next.run();
    } else {
      // maybe take a screenshot, second argument is true so we wait for i/o to complete
      screenShotMaybe(next, true);
    }
  }
```
This method promptly sets the currentView to `PROJECTS` hence the if condition does not fail and we get the expected behaviour.

In the remaining cases we avoid the bug as the initial view is designer which is same as assumed by Ode.

**Comment by Himanshu2107** [comment]
@ewpatton 
What can be the further course of action?

**Comment by ewpatton** [comment]
Your analysis seems pretty thorough. Please try making the change and if it works submit a Pull request.

**Comment by Himanshu2107** [comment]
For solving this bug I had made the following changes:
in com/google/appinventor/client/Ode.java

~~`public static int currentView = DESIGNER;`~~
`public static int currentView;`

and

inside `onModuleLoad()` in `onSuccess()` call of `OdeAsynCallBack` (lines 757 - 799):
```
userSettings = new UserSettings(user);
userSettings.loadSettings(new Command() { .......... }

// New code added here
// Set the current view of Ode
        if(getUserAutoloadProject()) {
          currentView = DESIGNER;
        } else {
          currentView = PROJECTS;
        }
```
It turns out that the call to `getUserAutoloadProject()` always return `true`. This is because the default value of `SettingsConstants.USER_AUTOLOAD_PROJECT` is set to `true` and it does not change on initializatoin (found in GeneralSettings.java).

@ewpatton What should I do to know whether `autoload` is on or off? I was not able to find another way. I will need your help in this.

I want to ask one more thing, should I ask these type of queries here or on the community website?

**Comment by Himanshu2107** [comment]
@ewpatton thoughts?

**Comment by ewpatton** [comment]
@Himanshu2107 It might just be simple enough to start with `currentView = PROJECTS`, which is the first view that is visible anyway. The fact that it is initialized to DESIGNER when in fact the first view is not the designer seems like the correct. This wouldn't require waiting for the user settings to be loaded.

**Comment by Himanshu2107** [comment]
Okay.
This change fixes both #2419 and #2433.

**Comment by ewpatton** [comment]
Fixed by #2436
